src/main/java/com/github/splendor_mobile_game/game/model/Game.java

```diff
@@ -343,4 +349,51 @@ private ArrayList<Noble> getRandomNobles(int amount){
     }
 
 
+
+
+    public void canPerformAnyAction(User user) throws CanPerformAnActionException {
+
+        // Check if user can reserve any card from deck
+        if (canReserveCardFromDeck(user, CardTier.LEVEL_1)) throw new CanPerformAnActionException("You can reserve a card of tier 1!");
+        if (canReserveCardFromDeck(user, CardTier.LEVEL_2)) throw new CanPerformAnActionException("You can reserve a card of tier 2!");
+        if (canReserveCardFromDeck(user, CardTier.LEVEL_3)) throw new CanPerformAnActionException("You can reserve a card of tier 3!");
+
+        // Check if user can buy or reserve card from table
+        for (Deck deck : revealedCards.values()) {
+            for (Card card : deck) {
+
+                // Check if user can reserve a card from table
+                if (canReserveCardFromTable(user, card))
+                    throw new CanPerformAnActionException(String.format("There is a card of %s tier which you can reserve!", deck.getTier().toString()));
+
+                // Check if user can buy a card from table
+                if (user.canBuyCard(card))
+                    throw new CanPerformAnActionException(String.format("You can buy a card of %s tier!", deck.getTier().toString()));
+
+            }
+        }
+
+
+        // Check if user can buy reserved card
+        for (Card card : user.getReservedCards())
+            if (user.canBuyCard(card)) throw new CanPerformAnActionException("You can buy one of your reserved cards!");
+
+
+
+        int greaterThanOneCount = 0;
+        // Check if user can take tokens
+        for (Map.Entry<TokenType, Integer> entry : tokensOnTable.entrySet()) {
+            if (entry.getKey() == TokenType.GOLD_JOKER) continue;
+
+            if (entry.getValue() >= 4)
+                throw new CanPerformAnActionException("You can take 2 tokens of any color!");
+
+            if (entry.getValue() >= 1) greaterThanOneCount += 1;
+        }
+
+        if (greaterThanOneCount >= 3)
+            throw new CanPerformAnActionException("You can take 3 tokens of different colors!");
```

Comment:
TLDR: to fix this we should fix the getting tokens logic first, because it has bugs as well. Checking for every possible scenario is such a pain.

Reply:
What are the unnecessary checks here? Maybe i can just leave TODO here and we will create an issue for GetTokens reaction's Author. So that he knows how to properly check the legalacy of the moves.

Reply:
I believe the original author has had enough. There is a need for a fresh perspective.